bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Bookie.java

```diff
@@ -1446,6 +1304,153 @@ public LedgerStorage getLedgerStorage() {
         return ledgerStorage;
     }
 
+    /**
+     * A implementation of StateManager.
+     */
+    public class BookieStateManager implements StateManager{
```

Comment:
- I think you can move this class out of Bookie. The current `Bookie` class is too huge. The tricky part is StateManager need to trigger bookie shutdown if it can't register to registration service. What you can do is adding a shutdown handler, bookie register a shutdown handler to state manager. if state manager can't not register, then it will trigger shutdown. An example code:

```

interface StateManager {

    interface ShutdownHandler {
         void shutdown();
    }

    void setShutdownHandler(ShutdownHandler handler);

}

```

- remove `public` to make it a package protected class
- need space before '{'

Once you moved this class out of `Bookie` class, you can write a unit test for this class. you can mock RegistrationManager. so we will have a good code coverage of testing bookie state transitions.